tee-worker/sidechain/rpc-handler/src/direct_top_pool_api.rs

```diff
@@ -333,27 +340,59 @@ where
 	response.map_err(|e| format!("{:?}", e))
 }
 
-async fn request_vc_inner(params: Params) -> Result<RpcReturnValue, String> {
+fn author_submit_request_vc_inner<R, TCS, G>(author: Arc<R>, params: Params) -> Result<H256, String>
+where
+	R: AuthorApi<H256, H256, TCS, G> + Send + Sync + 'static,
+	TCS: PartialEq + Encode + Decode + Debug + Send + Sync + 'static,
+	G: PartialEq + Encode + Decode + Debug + Send + Sync + 'static,
+{
 	let payload = get_request_payload(params)?;
-	let request = AesRequest::from_hex(&payload)
-		.map_err(|e| format!("AesRequest construction error: {:?}", e))?;
+	let request = AesRequest::from_hex(&payload).map_err(|e| format!("{:?}", e))?;
+
+	// Use this hash to trigger the storage of connection_registry
+	let hash = request.using_encoded(|x| BlakeTwo256::hash(x));
+
+	thread::spawn(move || request_vc_inner1(author, request, hash));
+
+	Ok(hash)
+}
 
+fn request_vc_inner1<R, TCS, G>(author: Arc<R>, request: AesRequest, hash: H256)
+where
+	R: AuthorApi<H256, H256, TCS, G> + Send + Sync + 'static,
+	TCS: PartialEq + Encode + Decode + Debug + Send + Sync + 'static,
+	G: PartialEq + Encode + Decode + Debug + Send + Sync + 'static,
+{
 	let vc_request_sender = VcRequestSender::new();
-	let (sender, receiver) = oneshot::channel::<Result<Vec<u8>, String>>();
+	let (sender, receiver) = channel::<Result<Vec<u8>, String>>();
 
-	vc_request_sender.send(VCRequest { sender, request })?;
+	if let Err(err) = vc_request_sender.send(VCRequest { sender, request }) {
+		error!("failed to send AesRequest within request_vc: {:?}", err);
+		return
+	}
 
-	// we only expect one response, hence no loop
-	match receiver.await {
-		Ok(Ok(response)) =>
-			Ok(RpcReturnValue { do_watch: false, value: response, status: DirectRequestStatus::Ok }),
-		Ok(Err(e)) => {
-			log::error!("Received error in jsonresponse: {:?} ", e);
-			Err(e)
-		},
-		Err(_) => {
-			// This case will only happen if the sender has been dropped
-			Err("The sender has been dropped".to_string())
-		},
+	let mut req_cnt = 0u8;
+	while let Ok(res) = receiver.recv() {
+		match res {
+			Ok(response) => {
+				req_cnt += 1;
+				if let Ok(vc_res) = RequestVcResultOrError::decode(&mut response.as_slice()) {
+					author.send_rpc_response(hash, response, req_cnt < vc_res.len);
+					if req_cnt >= vc_res.len {
+						break
```

Comment:
Does it actually close the ws connection?

If not, could we end up exhausting the connection limit quite soon?

Reply:
Yes. Once the receiver goes out of scope. The connection will be cut down automatically.

Reply:
It will.  Once `do_watch` is false, the wss connection will be cut down.